kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -374,6 +372,8 @@ private void forwardResponse(DecodedFrame<?, ?> decodedFrame, ResponseHeaderData
             }
             ctx.fireChannelRead(decodedFrame);
         }
+        // required to flush the message back to the client
+        ctx.fireChannelReadComplete();
```

Comment:
I agree this feels like the right change to make but I'm not sure it makes a practical difference due to the implementation of [deferredResponseCompleted](https://github.com/kroxylicious/kroxylicious/blob/ade31a6dca86014eb2622bdf012beacd09c18709/kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java#L310) which I think ensures the channel is flushed. 

Though if you have evidence (ideally a test case as @k-wall suggested) that this PR is needed then lets do it.

Reply:
For the short circuit request path, FilterIT#requestFiltersCanRespondWithoutProxying fails if fireChannelReadComplete is removed as the response never reaches the client.

For the normal response path we have io.kroxylicious.proxy.FilterIT#supportsForwardDeferred* which tests request and response forward where the result future is completed either

- by an outside thread doing the completion  (io.kroxylicious.proxy.filter.ForwardingStyle#ASYNCHRONOUS_DELAYED) or
- - the completion of a asynchronous request to the broker (io.kroxylicious.proxy.filter.ForwardingStyle#ASYNCHRONOUS_DELAYED_ON_EVENTlOOP).

I am curious what differs in @luozhenyu's case.

Reply:
Another piece of the puzzle to be aware of is that we switch back to the eventloop thread after async work has been done. So if we obtain a CompletionStage from the user's Filter implementation, and that Stage has not completed yet, then we chain it with [thenApplyAsync back on to the channels event loop](https://github.com/kroxylicious/kroxylicious/blob/51c1ade15df644009926c04a2e41e06a2a966dd5/kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java#L305).

This is done early, before `handleResponseFilterResult` (and `handleRequestFilterResult` in the other direction), so we should be on the event loop at that point.